fix #1052: Handle hallucinated tool name.

ehsavoie opened the pull request:
<!--
Thank you so much for your contribution!

Please fill in all the sections below.
Please open the PR as a draft initially. Once it is reviewed and approved, we will ask you to add documentation and examples.
Please note that PRs with breaking changes or without tests will be rejected.

Please note that PRs will be reviewed based on the priority of the issues they address.
We ask for your patience. We are doing our best to review your PR as quickly as possible.
Please refrain from pinging and asking when it will be reviewed. Thank you for understanding!
-->

## Issue
<!-- Please specify the ID of the issue this PR is addressing. For example: "Closes #1234" or "Fixes #1234" -->
Closes # 1052

## Change
If the tool name doesn't exist, LangChain4J shouldn't fail with a NPE. It should just ignore it.


## General checklist
<!-- Please double-check the following points and mark them like this: [X] -->
- [X] There are no breaking changes
- [x] I have added unit and/or integration tests for my change
- [ ] The tests cover both positive and negative cases
- [ ] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [ ] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green
<!-- Before adding documentation and example(s) (below), please wait until the PR is reviewed and approved. -->
- [ ] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)


## Checklist for adding new maven module
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have added my new module in the root `pom.xml` and `langchain4j-bom/pom.xml`


## Checklist for adding new embedding store integration
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreIT` that extends from either `EmbeddingStoreIT` or `EmbeddingStoreWithFilteringIT`
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreRemovalIT` that extends from `EmbeddingStoreWithRemovalIT`

## Checklist for changing existing embedding store integration
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have manually verified that the `{NameOfIntegration}EmbeddingStore` works correctly with the data persisted using the latest released version of LangChain4j

ehsavoie commented:
@langchain4j sorry I missed your comment on the issue. I'll try to think of something better

MarlonCarvalhoUO commented:
Hey! How are you? I hope so! Do you intend to create AI agents for specific tasks like crewAi and Langchain?

ehsavoie commented:
I've updated the code o that it will throw a RuntimeException in this case. This behavior can be changed using a ServiceLoader to provide for example an error message to send back to the LLM.

ehsavoie commented on langchain4j/src/main/java/dev/langchain4j/service/tool/HallucinatedToolExecutor.java:
```diff
@@ -0,0 +1,3 @@
+package dev.langchain4j.service.tool;
+
+public interface HallucinatedToolExecutor extends ToolExecutor {}
```
Removing it

ehsavoie commented on langchain4j/src/main/java/dev/langchain4j/service/DefaultAiServices.java:
```diff
@@ -50,25 +62,26 @@
         super(context);
+        HallucinatedToolExecutor loadedHallucinatedToolExecutor = null;
+        for (HallucinatedToolExecutorFactory factory :
+                ServiceHelper.loadFactories(HallucinatedToolExecutorFactory.class)) {
```
Ok, since guardrails should protect against that I'll keep the code simple.

ehsavoie commented:
@langchain4j I've simplified the code to throw an exception and even added a test ;)

ehsavoie committed:
Adding support for invalid tool call from the LLM.
Currently the call will result in a RuntimeException with a meaningful message.

Signed-off-by: Emmanuel Hugonnet <ehugonne@redhat.com>

dliubarskyi commented:
Hi @ehsavoie, thanks a lot for the initial implementation!

@mariofusco has completed it in the https://github.com/langchain4j/langchain4j/pull/2460, so I am closing this issue
